metrics/ci_integrations/lib/source/buildkite/adapter/buildkite_source_client_adapter.dart

```diff
@@ -0,0 +1,247 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:ci_integration/client/buildkite/buildkite_client.dart';
+import 'package:ci_integration/client/buildkite/models/buildkite_artifact.dart';
+import 'package:ci_integration/client/buildkite/models/buildkite_artifacts_page.dart';
+import 'package:ci_integration/client/buildkite/models/buildkite_build.dart';
+import 'package:ci_integration/client/buildkite/models/buildkite_build_state.dart';
+import 'package:ci_integration/client/buildkite/models/buildkite_builds_page.dart';
+import 'package:ci_integration/integration/interface/source/client/source_client.dart';
+import 'package:ci_integration/util/model/interaction_result.dart';
+import 'package:meta/meta.dart';
+import 'package:metrics_core/metrics_core.dart';
+
+/// An adapter for [BuildkiteClient] to implement the [SourceClient]
+/// interface.
+class BuildkiteSourceClientAdapter implements SourceClient {
+  /// A fetch limit for Buildkite API calls.
+  static const int fetchLimit = 25;
+
+  /// A [BuildkiteClient] instance to perform API calls.
+  final BuildkiteClient buildkiteClient;
+
+  /// Creates a new instance of the [BuildkiteSourceClientAdapter].
+  ///
+  /// Throws an [ArgumentError] if [buildkiteClient] is `null.
+  BuildkiteSourceClientAdapter({
+    @required this.buildkiteClient,
+  }) {
+    ArgumentError.checkNotNull(buildkiteClient, 'buildkiteClient');
+  }
+
+  @override
+  Future<List<BuildData>> fetchBuilds(String pipeline) async {
+    return _fetchLatestBuilds(pipeline);
+  }
+
+  @override
+  Future<List<BuildData>> fetchBuildsAfter(
+    String pipeline,
+    BuildData build,
+  ) async {
+    ArgumentError.checkNotNull(build, 'build');
+    final latestBuildNumber = build.buildNumber;
+
+    final firstBuildsPage = await _fetchBuildsPage(
+      pipeline,
+      page: 1,
+      perPage: 1,
+    );
+
+    final builds = firstBuildsPage.values ?? [];
+
+    if (builds.isEmpty || builds.first.number <= latestBuildNumber) return [];
+
+    return _fetchLatestBuilds(pipeline, latestBuildNumber);
+  }
+
+  /// Fetches the latest builds by the given [pipeline].
+  ///
+  /// If the [latestBuildNumber] is not `null`, returns all builds with the
+  /// [Build.buildNumber] greater than the given [latestBuildNumber]. Otherwise,
+  /// returns no more than the latest [fetchLimit] builds.
+  Future<List<BuildData>> _fetchLatestBuilds(
+    String pipeline, [
+    int latestBuildNumber,
+  ]) async {
+    final List<BuildData> result = [];
+    bool hasNext = true;
+
```

Comment:
The Buildkite API allows filtering builds with the `finished_from` parameter as listed [here](https://buildkite.com/docs/apis/rest-api/builds#list-builds-for-a-pipeline). Moreover, we've added this to the client implementation. Should we use this filtering in an adapter as it seems to be much better for builds fetching?